p2p/starknet/handlers.go

```diff
@@ -134,16 +135,47 @@ func (h *Handler) onBlockBodiesRequest(req *spec.BlockBodiesRequest) (Stream[pro
 }
 
 func (h *Handler) onEventsRequest(req *spec.EventsRequest) (Stream[proto.Message], error) {
-	// todo: read from bcReader and adapt to p2p type
-	count := uint64(0)
-	return func() (proto.Message, bool) {
-		if count > 3 {
+	it, err := h.newIterator(req.Iteration)
+	if err != nil {
+		return nil, err
+	}
+
+	var finSent bool
+	fin := func() (proto.Message, bool) {
+		if finSent {
 			return nil, false
 		}
-		count++
+		finSent = true
+
 		return &spec.EventsResponse{
-			Id: &spec.BlockID{
-				Number: count - 1,
+			Responses: &spec.EventsResponse_Fin{},
+		}, true
+	}
+
+	return func() (proto.Message, bool) {
+		if !it.Valid() {
+			return fin()
+		}
+
+		block, err := it.Block()
+		if err != nil {
+			h.log.Errorw("Failed to fetch block", "err", err)
+			return fin()
+		}
+		it.Next()
+
+		items := make([]*spec.Event, 0, len(block.Receipts))
+		for _, item := range block.Receipts {
+			events := utils.Map(item.Events, core2p2p.AdaptEvent)
```

Comment:
this is another allocation we can avoid by simply looping over item.Events and appending each event to `items` one by one after adapting.